wordpress-mobile/WordPress-FluxC-Android
# Use a sealed class for cause of change for OnPostChanged

We don't have any information about which post might be changed in the `OnPostChanged` event and it has caused some not-so-ideal implementations before such as [this](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java#L732-L739). Unfortunately, that kind of implementation would cause too big of a problem with the new `ListStore` implementation, since all posts will be fetched individually. (for now) So, in this PR I am trying to address this issue.

I believe we should use different events for different actions and get rid of `causeOfChange` completely, but then again, there is a lot of improvements we should be making in `PostStore` and I don't think making major changes as part of another project is the correct way to go. So, I am going for a minor improvement with the least amount of changes possible in the client.

This PR adds a new sealed class `CauseOfOnPostChanged` to be used for `OnPostChanged.causeOfChanged`. I have added `localPostId` and `remotePostId` to the classes that made sense. Please note that these properties might be `0` in some cases. It's just not possible to set that information without functional changes in those cases. However, we can always improve on them at a later time with this refactor out of the way.

Rest of the changes are one to one replacement of old `causeOfChange` with the new one.

To test:

1. Run tests in `ReleaseStack_PostTestWPCom.java`, `ReleaseStack_PostTestXMLRPC.java`, `PageStoreTest.kt`. (Please note that although rare I find some of these tests can fail unrelated to these changes, this has always been the case. Make sure to run the tests a few times if it fails.)
2. Run the example app and Instaflux and make sure they build OK.
3. Check that fetching posts/pages work fine in the example app (log shown)
4. Make sure all `OnPostChanged` events are updated. I checked each one, but it'd be good if the reviewer can check as well to make sure I didn't miss anything. Unfortunately compiler is not helpful for catching this change since Java can compare two different types... 🤦‍♂️

The WPAndroid PR is almost ready and will be mentioned as a comment when it's up. It'd be great to review that PR as well before merging this PR.

@0nko Can I bother you with another review as these changes are touching some page logic and it looks like you worked on these classes recently, both in FluxC and WPAndroid (at least that's what git blame tells me)?
